### ISSUE TITLE
`unbuffered_bytes` lint does not provide a better solution

### ISSUE THREAD

**Issue by schultetwin1** [rust-lang/rust-clippy#14340]
### Summary

My [nightly build](https://github.com/schultetwin1/debpkg/actions/runs/13621783862) of https://github.com/schultetwin1/debpkg/ is failing due to the `unbuffered_bytes` lint.

```
error: calling .bytes() is very inefficient when data is not in memory
  --> src/debian_binary.rs:21:17
   |
21 |     for byte in stream.bytes() {
   |                 ^^^^^^^^^^^^^^
   |
   = help: consider using `BufReader`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unbuffered_bytes
   = note: `-D clippy::unbuffered-bytes` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::unbuffered_bytes)]`

error: could not compile `debpkg` (lib) due to 1 previous error
Error: Process completed with exit code 101.
```

`debpkg` is a library crate to be used to parse debian packages. Users are free to use a `BufReader` if they desired but I do not want to force users to use a `BufReader`.

As a crate author asking for a type which impls`Reader` in my public API and also using the `bytes()` function what should I do? Am I missing a better option here?

Playground Reproduction: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=d30f695dde2d2ab492de1cdeeb47fe57

### Reproducer

```rust
use std::io::Read;

#[derive(Debug)]
pub struct DebianBinaryVersion {
    pub major: u32,
    pub minor: u32,
}

pub fn parse_debian_binary_contents<R: Read>(stream: &mut R) -> Result<DebianBinaryVersion, &'static str> {
    let mut first_two_bytes: [u8; 2] = [0, 0];
    stream.read_exact(&mut first_two_bytes).map_err(|_| "Read Failed")?;

    if &first_two_bytes != b"2." {
        return Err("Invalid Version");
    }

    // note: This limits the largest minor version to 99999. Hopefully we never get above that.
    let mut string = String::new();
    for byte in stream.bytes() {
        let byte = byte.map_err(|_| "Read of byte failed")?;
        if byte == b'\n' {
            break;
        }
        if !(byte as char).is_ascii_digit() {
            return Err("Version does not contain ASCII");
        }
        string.push(byte as char);
    }
    let minor = match string.as_str().parse::<u32>() {
        Ok(num) => num,
        Err(_e) => return Err("Error is not a number"),
    };

    Ok(DebianBinaryVersion { major: 2, minor })
}
```

### Version

```text
rustc 1.87.0-nightly (f4a216d28 2025-03-02)
binary: rustc
commit-hash: f4a216d28ee635afce685b4206e713579f66e130
commit-date: 2025-03-02
host: aarch64-apple-darwin
release: 1.87.0-nightly
LLVM version: 20.1.0
```

**Comment by samueltardieu** [comment]
If you don't want to force your users to use a `BufReader` in your public API, you can:

- add an `#[expect(clippy::unbuffered_bytes)]` to your method (use `allow` if you have an older MSRV)
- add a note to your method documentation suggesting that the user wrap their stream into a `BufReader` to avoid getting bytes one by one because this may be inefficient

If the lint helps you write a better documentation for your users, I would say it has fulfilled its usage.
